lib/tpm2_errata.c

```diff
@@ -0,0 +1,212 @@
+
+    struct tpm2_errata_desc *errata;
+    va_list ap;
+
+    if (errata_query(nr, &errata) == false)
+        return false;
+
+    va_start(ap, nr);
+    errata->fixup(errata, &ap);
+    va_end(ap);
+
+    return true;
+}
+
+const char *tpm2_errata_name(int nr) {
+
+    struct tpm2_errata_desc *errata;
+
+    if (errata_query(nr, &errata) == false)
+        return false;
+
+    return errata->name;
+}
+
+void tpm2_errata_init(TSS2_SYS_CONTEXT *sapi_ctx) {
+
+    TPMS_CAPABILITY_DATA      capability_data;
+    TPML_TAGGED_TPM_PROPERTY *properties;
+    TSS2_RC                   rc;
+    TPMI_YES_NO               more_data;
+    size_t                    i;
+    UINT32                    spec_rev, day_of_year, year;
+
+    rc = Tss2_Sys_GetCapability(sapi_ctx,
+                                NULL,
+                                TPM_CAP_TPM_PROPERTIES,
+                                PT_FIXED,
+                                MAX_TPM_PROPERTIES,
+                                &more_data,
+                                &capability_data,
+                                NULL);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERR("Failed to GetCapability: capability: 0x%x, property: 0x%x, "
+                "TSS2_RC: 0x%x\n", TPM_CAP_TPM_PROPERTIES, PT_FIXED, rc);
+        return;
+    } else if (more_data == YES)
+        LOG_WARN("More data to be queried: capability: 0x%x, property: "
+                 "0x%x\n", TPM_CAP_TPM_PROPERTIES, PT_FIXED);
+
+    spec_rev = day_of_year = year = 0;
+    properties = &capability_data.data.tpmProperties;
+    for (i = 0; i < properties->count; ++i) {
+        TPMS_TAGGED_PROPERTY *property = properties->tpmProperty + i;
+
+        if (property->property == TPM_PT_REVISION)
+            spec_rev = property->value;
+        else if (property->property == TPM_PT_DAY_OF_YEAR)
+            day_of_year = property->value;
+        else if (property->property == TPM_PT_YEAR)
+            year = property->value;
+        else if (property->property > TPM_PT_YEAR)
+            break;
+    }
+
+    if (!spec_rev || !day_of_year || !year) {
+        LOG_ERR("Invalid TPM_SPEC parameter\n");
+        return;
+    }
+
+    /* Determine the TPM spec and errata. */
+    for (i = 0; i < sizeof(known_errata_info) /
+                    sizeof(struct tpm2_errata_info); ++i) {
+         if (known_errata_info[i].year == year &&
+             known_errata_info[i].day_of_year == day_of_year &&
+             known_errata_info[i].spec_rev == spec_rev) {
+             this_errata_info = known_errata_info + i;
+             break;
+         }
+    }
+
+    if (!this_errata_info) {
+        LOG_ERR("Unknow TPM_SPEC. spec_rev: 0x%x, year: %d, day_of_year: %d\n",
+                spec_rev, year, day_of_year);
+        return;
+    }
+
+    LOG_INFO("TPM_SPEC: spec rev %f, errata ver %f\n",
+             (float)this_errata_info->spec_rev / 100,
+             (float)this_errata_info->errata_ver / 100);
+}
```

Comment:
Failures in this routine are fatal at errata lookup, which fails if initialization is not performed... Should looking up the tpm version fail in the tools?